### PR TITLE
Feature/v1-189 | Passage time issue.

### PR DESCRIPTION
**Pull request by dapodev**
# Changes

- Backend
1. Added `startTestDate` and `finishTestDate` fields to models, remove `testDate` field.
1. Added ability to calculate `finishTestDate` as estimated finish date when starting the test.
1. Upgraded check dates utilities.
1. Updating status feature when attemting to getTest after test timeout expired.
1. Test date validation among controllers

- Frontend
1. Added one timeout value for test in different tabs
1. Refactored logic with redirect in PassingTest page

### CODE REVIEW

File: frontend/src/pages/PassingTest/PassingTestContainer.tsx [hunk 1]
```diff
@@ -95,9 +104,18 @@ const PassingTestContainer: React.FC = () => {
   const questionStageItem = courseTest?.questions[stage - 1];
   const courseStatus = clientCourseResponse?.status;
 
-  const isShouldRedirect =
+  const isTestTimeOver = useMemo(
+    () =>
+      courseStatus === COURSE_STATUSES.testing &&
+      getDurationBetweenDates(clientCourseResponse?.finishTestDate) < 0,
```
Comment (irina-sinitsyna) [inline frontend/src/pages/PassingTest/PassingTestContainer.tsx:110]: please replace magic number with constant

File: frontend/src/pages/PassingTest/PassingTestContainer.tsx [hunk 2]
```diff
@@ -116,17 +134,21 @@ const PassingTestContainer: React.FC = () => {
...
-  useEffect(() => {
-    if (courseTest) {
-      setShowDialogOnSwitchingRoute(true);
-    }
-  }, [courseTest, showDialogOnSwitchingRoute]);
+  // useEffect(() => {
+  //   if (courseTest) {
+  //     setShowDialogOnSwitchingRoute(true);
+  //   }
+  // }, [courseTest, showDialogOnSwitchingRoute]);
```
Comment (irina-sinitsyna) [inline frontend/src/pages/PassingTest/PassingTestContainer.tsx:147]: Let's remove comments?

File: frontend/src/pages/PassingTest/PassingTestContainer.tsx (hunk 1 above)
Comment (ostapdans99) [reply frontend/src/pages/PassingTest/PassingTestContainer.tsx]: done

File: frontend/src/pages/PassingTest/PassingTestContainer.tsx (hunk 2 above)
Comment (ostapdans99) [reply frontend/src/pages/PassingTest/PassingTestContainer.tsx]: done

**Review by irina-sinitsyna — APPROVED** [review]

File: backend/src/interfaces/Ientities/IclientCourses.d.ts [hunk 3]
```diff
@@ -33,7 +33,8 @@ interface IClientCourse {
   ];
   date: Date;
   applyDate?: Date;
-  testDate?: Date;
```
Comment (SaXaPhonist) [inline backend/src/interfaces/Ientities/IclientCourses.d.ts:36]: if you replace testDate please also replace all connected data and fns with testDate on frontend  according to new logic

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

File: backend/src/interfaces/Ientities/IclientCourses.d.ts (hunk 3 above)
Comment (ostapdans99) [reply backend/src/interfaces/Ientities/IclientCourses.d.ts:28]: done

**Pull request merged**